Zend/zend_attributes.c

```diff
@@ -123,34 +235,70 @@ ZEND_API zend_attribute *zend_add_attribute(HashTable **attributes, zend_bool pe
 	return attr;
 }
 
-ZEND_API void zend_compiler_attribute_register(zend_class_entry *ce, zend_attributes_internal_validator validator)
+static void free_internal_attribute(zval *v)
 {
+	pefree(Z_PTR_P(v), 1);
+}
+
+ZEND_API zend_internal_attribute *zend_internal_attribute_register(zend_class_entry *ce, uint32_t flags)
+{
+	zend_internal_attribute *attr;
+
 	if (ce->type != ZEND_INTERNAL_CLASS) {
 		zend_error_noreturn(E_ERROR, "Only internal classes can be registered as compiler attribute");
 	}
 
+	attr = pemalloc(sizeof(zend_internal_attribute), 1);
+	attr->ce = ce;
+	attr->flags = flags;
+	attr->validator = NULL;
+
 	zend_string *lcname = zend_string_tolower_ex(ce->name, 1);
 
-	zend_hash_update_ptr(&internal_validators, lcname, validator);
+	zend_hash_update_ptr(&internal_attributes, lcname, attr);
+	zend_add_class_attribute(ce, zend_ce_attribute->name, 0);
 	zend_string_release(lcname);
 
-	zend_add_class_attribute(ce, zend_ce_php_attribute->name, 0);
+	return attr;
 }
 
-void zend_register_attribute_ce(void)
+ZEND_API zend_internal_attribute *zend_internal_attribute_get(zend_string *lcname)
 {
-	zend_hash_init(&internal_validators, 8, NULL, NULL, 1);
+	return (zend_internal_attribute *) zend_hash_find_ptr(&internal_attributes, lcname);
```

Comment:
```suggestion
	return zend_hash_find_ptr(&internal_attributes, lcname);
```

These aren't needed in C.